src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,168 @@
+import duke.exceptions.InvalidDateTimeException;
+import duke.exceptions.InvalidTaskSpecificationException;
+import duke.exceptions.MissingArgumentException;
+import duke.exceptions.NoCommandException;
+import duke.exceptions.UnknownCommandException;
+import duke.parser.Parser;
+import duke.storage.Storage;
+import duke.tasklist.TaskList;
+import duke.ui.Ui;
+
+/**
+ * The type duke.Duke.
+ */
+public class Duke {
+
+    private static final String defaultFilePath = "src/main/java/duke/data/tasks.txt";
+    private final Storage storage;
+    private final TaskList tasks;
+    private final Ui ui;
+    private final Parser parser;
+    private final String filepath;
+
+    /**
+     * Duke Constructor
+     *
+     * @param filePath
+     *            The path to the file where the data is stored.
+     */
+    public Duke(String filePath) {
+        this.filepath = filePath;
+        ui = new Ui();
+        storage = new Storage();
+        tasks = new TaskList(storage.load(this.filepath));
+        parser = new Parser();
+    }
+
+    /**
+     * Duke Constructor
+     */
+    public Duke() {
+        this.filepath = defaultFilePath;
+        ui = new Ui();
+        storage = new Storage();
+        tasks = new TaskList(storage.load(this.filepath));
+        parser = new Parser();
+    }
+
+    /**
+     * The runCommandLoopUntilExitCommand function runs a loop that will continue to
+     * run until the user enters an exit command.
+     * It first asks for a command from the user, and then tries to parse it using
+     * the Parser class. If there is an error in parsing,
+     * it will display that error message and ask for another command. Otherwise, if
+     * there are no errors in parsing, it runs runCommand on
+     * the parsed Command object (which returns a CommandResult). The function then
+     * displays this result on screen using Ui's showResult method.
+     */
+    private void runCommandLoopUntilExitCommand() {
+        BaseCommand command;
+        do {
+            String userCommandText = ui.getUserCommand();
+            try {
+                command = parser.parse(userCommandText);
+            } catch (MissingArgumentException | InvalidDateTimeException | InvalidTaskSpecificationException
+                    | IncorrectArgumentException | UnknownCommandException | NoCommandException e) {
+                command = new ErrorCommand(e.getMessage());
+            }
+            CommandResult result = runCommand(command);
+            ui.showResult(result);
+        } while (!ExitCommand.equals(command));
+    }
+
+    /**
+     * The runCommand function takes a BaseCommand object as an argument and
+     * executes it.
+     * If the command is a BaseTaskCommand, then the task list is set to this
+     * TaskList object.
+     * The function returns the result of executing that command.
+     *
+     * @param command
+     *            Pass the command object to the run command method
```

Comment:
This line of description should be in line with @param command, otherwise the indentation may look off. :)

Reply:
Ah thank you, I will take note of this!